### ISSUE TITLE
libssl.so.1.1 & libcrypto.so.1.1

### ISSUE THREAD

**Issue by shoujii** [openssl/openssl#5395]
Hey there,

maybe it's a dumb question, but I'm receiving this error, after I compiled the 1.1.1pre1 Version.

```
root@:/usr/local# uname -a
Linux 4.9.0-5-amd64 #1 SMP Debian 4.9.65-3+deb9u2 (2018-01-04) x86_64 GNU/Linux
```

compile options:
```
./config --prefix=/usr --openssldir=/etc/ssl --libdir=lib shared zlib-dynamic
```

```
root@:/usr/local# which openssl
/usr/bin/openssl
```

```
root@:/usr/local# ldd /usr/bin/openssl
/usr/bin/openssl: /usr/lib/x86_64-linux-gnu/libssl.so.1.1: version `OPENSSL_1_1_1' not found (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1: version `OPENSSL_1_1_1' not found (required by /usr/bin/openssl)
        linux-vdso.so.1 (0x00007ffcae361000)
        libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f6303c4f000)
        libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f63037bc000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f63035b8000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f630339b000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6302ffc000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f630416e000)
```

regards,
Shoujii

**Comment by noloader** [comment]
You have to add additional `CFLAGS` when building things so the libraries can be found. For some reason, folks think the dominant use case is, build using a particular version of `libssl.so` and `libcrypto.so`, but dont't use them at runtime. Derp...

Change this:

```
./config --prefix=/usr --openssldir=/etc/ssl --libdir=lib shared zlib-dynamic
```

To:

```
./config --prefix=/usr --openssldir=/etc/ssl --libdir=lib shared zlib-dynamic -Wl,-R,/etc/ssl/lib -Wl,--enable-new-dtags
```

There are other bug reports about the same problem. It is probably not going to be fixed.

**Comment by shoujii** [comment]
Hey, thanks for the reply..

I compiled Openssl with that /config options, but still the same error.

```
root@:/home/openssl-1.1.1-pre1# ldd /usr/bin/openssl
/usr/bin/openssl: /usr/lib/x86_64-linux-gnu/libssl.so.1.1: version `OPENSSL_1_1_1' not found (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1: version `OPENSSL_1_1_1' not found (required by /usr/bin/openssl)
        linux-vdso.so.1 (0x00007ffe9e704000)
        libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f5443470000)
        libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f5442fdd000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f5442dd9000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f5442bbc000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f544281d000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f544398f000)
```

**Comment by levitte** [comment]
> ```
> ./config --prefix=/usr --openssldir=/etc/ssl --libdir=lib shared zlib-dynamic -Wl,-R,/etc/ssl/lib -Wl,--enable-new-dtags
> ```

@noloader got it wrong, the rpath should be `/usr/lib`.
To avoid having to guess, do this instead:

```
./config --prefix=/usr --openssldir=/etc/ssl --libdir=lib shared zlib-dynamic -Wl,-R,'$(LIBRPATH)' -Wl,--enable-new-dtags
```

`$(LIBRPATH)` is a convenience variable in the Makefile, made for exactly this purpose. @noloader knows this.

**Comment by noloader** [comment]
> @noloader got it wrong, the rpath should be /usr/lib.

My bad, it looked like he was using `/etc/ssl` as a prefix.

**Comment by shoujii** [comment]
Thanks for the reply @levitte !

`./config --prefix=/usr --openssldir=/etc/ssl --libdir=lib shared zlib-dynamic -Wl,-R,'$(LIBRPATH)' -Wl,--enable-new-dtags`

solved my problem with compiling the 1.1.1pre1.
And I'm sorry, I wasn't aware of that $(LIBRPATH) variable (I thought that --libdir=lib is doing the job).
